core/primitives/address.hpp

```diff
@@ -3,16 +3,146 @@
+  };
+
+  enum class Network : int { testnet = 0, mainnet };
+
+  enum class Protocol : int { ID = 0, SECP256K1, Actor, BLS };
+
+  template <typename T>
+  constexpr auto to_underlying(T v) noexcept {
+    return static_cast<std::underlying_type_t<T>>(v);
+  }
+
+  struct PublicKeySecp256k1Hash {
+    PublicKeySecp256k1Hash(const Bytes &h) : hash(h) {}
+    PublicKeySecp256k1Hash(const Bytes &&h) : hash(h) {}
+
+    Bytes hash;
+    static const size_t size = 20;
+  };
+
+  struct ActorExecHash {
+    ActorExecHash(const Bytes &h) : hash(h) {}
+    ActorExecHash(const Bytes &&h) : hash(h) {}
+
+    Bytes hash;
+    static const size_t size = 20;
+  };
+
+  struct PublicKeyBLSHash {
+    PublicKeyBLSHash(const Bytes &h) : hash(h) {}
+    PublicKeyBLSHash(const Bytes &&h) : hash(h) {}
+
+    Bytes hash;
+    static const size_t size = 48;
+  };
+
+  bool operator==(const PublicKeySecp256k1Hash &lhs,
+                  const PublicKeySecp256k1Hash &rhs);
+  bool operator==(const ActorExecHash &lhs, const ActorExecHash &rhs);
+  bool operator==(const PublicKeyBLSHash &lhs, const PublicKeyBLSHash &rhs);
 
+  /**
+   * @brief Address interface declaration
+   */
   class Address {
+    using Payload = boost::
+        variant<Uint, PublicKeySecp256k1Hash, ActorExecHash, PublicKeyBLSHash>;
+
+   public:
+    // Creates an empty Address
+    Address() : networkId(Network::testnet), data(Uint{0}), empty(true) {}
+
+   private:
+    Address(Network net, Uint id) : networkId(net), data(id), empty(false) {}
+    Address(Network net, PublicKeySecp256k1Hash &&id)
+        : networkId(net), data(id), empty(false) {}
+    Address(Network net, ActorExecHash &&id)
+        : networkId(net), data(id), empty(false) {}
+    Address(Network net, PublicKeyBLSHash &&id)
+        : networkId(net), data(id), empty(false) {}
+
+    Bytes toBytes();
+
    public:
-    //TODO (kamilsa): FIL-21 Implement primitives
+    // Factory method to create Address object
+    static outcome::result<std::unique_ptr<Address>> makeAddress(
+        Network net, Byte protocol, const Bytes &payload);
+
+    // TODO: Implement creation of an Address by decoding a String (if
+    // necessary) static outcome::result< std::unique_ptr<Address> >
+    // decode(String s);
+
+    const Network getNetwork() {
+      return networkId;
+    }
+    const bool isEmpty() {
+      return empty;
+    }
+    const Payload &getData() {
+      return data;
+    };
+    Protocol getProtocol();
+
+    Bytes checksum();
+
+    // New API defined in
+    // https://filecoin-project.github.io/specs/#systems__filecoin_vm__actor__address
+    bool verifySyntax();
+    bool isKeyType();
+    bool equals(Address &other);
+    String string();
+
+    // Older API defined in
+    // https://filecoin-project.github.io/specs/#appendix__address Should it be
+    // implemented, as well?
+    String encode() {
+      return this->string();
+    }
+    bool validateChecksum(Bytes checksum) {
+      return true;
+    }
```

Comment:
Can you add operator <, please? So, it van be used in maps.
Operator == also will be helpful.

Reply:
Bytes is an alias for std::vector<uint8_t>, therefore the == operator comes for free. Noted on < though

Reply:
I mean for Address class.

Reply:
Thanks a lot!